lib/buffer.js

```diff
@@ -176,12 +176,27 @@ Buffer.from = function(value, encodingOrOffset, length) {
   if (isAnyArrayBuffer(value))
     return fromArrayBuffer(value, encodingOrOffset, length);
 
+  if (value == null)
+    throw new TypeError(kFromErrorMsg);
+
+  if (typeof value === 'number')
+    throw new TypeError('"value" argument must not be a number');
+
+  const valueOf = value.valueOf && value.valueOf();
+  if (valueOf != null && valueOf !== value)
+    return Buffer.from(valueOf, encodingOrOffset, length);
+
   var b = fromObject(value);
   if (b)
     return b;
 
-  if (typeof value === 'number')
-    throw new TypeError('"value" argument must not be a number');
+  if (typeof value === 'object' &&
```

Comment:
tiny nit: you can try to access properties on any `!(value === undefined || value === null)` value so maybe just `(typeof value[Symbol.toPrimitive] === 'function')`?